### PR TITLE
Spring Boot Data JPA

### PR DESCRIPTION
**Pull request by 4Vitalii5**
Requirements
Replace the existing BookRepository with the one that will use JpaRepository interface
Add Liquibase support
Change spring.jpa.hibernate.ddl-auto property value to validate
Implement soft delete concept for the Book entity
Add missing endpoints and methods for BookService
As a result of this HW you should have the following endpoints in the BookController:

Book Endpoints:

GET: /api/books (Retrieve book catalog) (should be done in the previous PR)
GET: /api/books/{id} (Retrieve book details) (should be done in the previous PR)
POST: /api/books (Create a new book) (should be done in the previous PR)
PUT: /api/books/{id} (Update a specific book)
Example of request body:

{
  "title": "Updated Title",
  "author": "Updated Author",
  "isbn": "978-1234567890",
  "price": 19.99,
  "description": "Updated description",
  "coverImage": "https://example.com/updated-cover-image.jpg"
}

DELETE /api/books/{id} (Delete a specific book)

### CODE REVIEW

File: src/main/java/mate/academy/model/Book.java [hunk 1]
```diff
@@ -29,5 +33,8 @@ public class Book {
     @NonNull
     private BigDecimal price;
     private String description;
+    @Column(name = "cover_image")
     private String coverImage;
+    @Column(name = "is_deleted", nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/model/Book.java]:
```suggestion
    private String coverImage;
    @Column(nullable = false)
```

File: src/main/java/mate/academy/service/impl/BookServiceImpl.java [hunk 2]
```diff
@@ -4,22 +4,29 @@
...
 @RequiredArgsConstructor
 public class BookServiceImpl implements BookService {
     private final BookRepository bookRepository;
     private final BookMapper bookMapper;
 
+    @Transactional
     @Override
     public BookDto save(CreateBookRequestDto requestDto) {
         Book book = bookMapper.toModel(requestDto);
+        if (bookRepository.existsByIsbn(book.getIsbn())) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/service/impl/BookServiceImpl.java:24]: remove checking

File: src/main/java/mate/academy/exception/DuplicateResourceException.java [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package mate.academy.exception;
+
+public class DuplicateResourceException extends RuntimeException {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/exception/DuplicateResourceException.java:3]: remove

File: src/main/java/mate/academy/service/impl/BookServiceImpl.java [hunk 4]
```diff
@@ -33,9 +40,35 @@ public List<BookDto> findAll() {
...
     public BookDto findById(Long id) {
-        Book book = bookRepository.findById(id).orElseThrow(() ->
+        Book book = getBookById(id);
+        return bookMapper.toDto(book);
+    }
+
+    @Override
+    public BookDto updateBookById(Long id, CreateBookRequestDto requestDto) {
+        Book book = getBookById(id);
+        if (!book.getIsbn().equals(requestDto.isbn())) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/service/impl/BookServiceImpl.java]: remove checking

File: src/main/java/mate/academy/service/impl/BookServiceImpl.java [hunk 5]
```diff
@@ -33,9 +40,35 @@ public List<BookDto> findAll() {
...
+            throw new DuplicateResourceException("Book with ISBN " + requestDto.isbn()
+                    + " already exists");
+        }
+        book.setTitle(requestDto.title());
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/service/impl/BookServiceImpl.java]: user mapper for these logic

File: src/main/java/mate/academy/service/impl/BookServiceImpl.java [hunk 6]
```diff
@@ -33,9 +40,35 @@ public List<BookDto> findAll() {
...
+        book.setPrice(requestDto.price());
+        book.setIsbn(requestDto.isbn());
+        book.setDescription(requestDto.description());
+        book.setCoverImage(requestDto.coverImage());
+        return bookMapper.toDto(bookRepository.save(book));
+    }
+
+    @Override
+    public void deleteById(Long id) {
+        getBookById(id);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/service/impl/BookServiceImpl.java]: redundant query to DB

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/mate/academy/service/impl/BookServiceImpl.java (hunk 2 above)
Comment (4Vitalii5) [reply src/main/java/mate/academy/service/impl/BookServiceImpl.java:24]: I removed it, but when I tried to add a book with an invalid ISBN DB increments ID and when I used a valid ISBN, DB it use already id+1. What should I do in this case?

File: src/main/java/mate/academy/service/impl/BookServiceImpl.java [hunk 7]
```diff
@@ -10,13 +10,15 @@
...
 import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
 
 @Service
 @RequiredArgsConstructor
 public class BookServiceImpl implements BookService {
     private final BookRepository bookRepository;
     private final BookMapper bookMapper;
 
+    @Transactional
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/service/impl/BookServiceImpl.java]:
```suggestion
```

File: src/main/java/mate/academy/service/impl/BookServiceImpl.java [hunk 8]
```diff
@@ -33,9 +35,31 @@ public List<BookDto> findAll() {
...
+    @Override
+    public BookDto updateBookById(Long id, CreateBookRequestDto requestDto) {
+        Book book = getBookById(id);
+        Book updatedBook = bookMapper.toModel(requestDto);
+        book.setTitle(updatedBook.getTitle());
+        book.setAuthor(updatedBook.getAuthor());
+        book.setPrice(updatedBook.getPrice());
+        book.setIsbn(updatedBook.getIsbn());
+        book.setDescription(updatedBook.getDescription());
+        book.setCoverImage(updatedBook.getCoverImage());
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/service/impl/BookServiceImpl.java]:
```suggestion
        bookMapper.updateBookFromDto(requestDto, book);
```

File: src/main/java/mate/academy/repository/BookRepository.java [hunk 9]
```diff
@@ -1,13 +1,8 @@
...
+import org.springframework.data.jpa.repository.JpaRepository;
 
-public interface BookRepository {
-    Book save(Book book);
-
-    List<Book> findAll();
-
-    Optional<Book> findById(Long id);
+public interface BookRepository extends JpaRepository<Book, Long> {
+    boolean existsByIsbn(String isbn);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/repository/BookRepository.java:7]: Check if you use this method, if no - remove it

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/mate/academy/mapper/BookMapper.java [hunk 10]
```diff
@@ -5,10 +5,20 @@
...
     BookDto toDto(Book book);
 
     Book toModel(CreateBookRequestDto requestDto);
+
+    @Mapping(source = "title", target = "title")
+    @Mapping(source = "author", target = "author")
+    @Mapping(source = "isbn", target = "isbn")
+    @Mapping(source = "price", target = "price")
+    @Mapping(source = "description", target = "description")
+    @Mapping(source = "coverImage", target = "coverImage")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/mapper/BookMapper.java]:
```suggestion
```

File: src/main/java/mate/academy/mapper/BookMapper.java [hunk 11]
```diff
@@ -5,10 +5,20 @@
...
+    Book updateBookFromDto(CreateBookRequestDto requestDto, @MappingTarget Book book);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/mapper/BookMapper.java]:
```suggestion
    void updateBookFromDto(CreateBookRequestDto requestDto, @MappingTarget Book book);
```

File: src/main/java/mate/academy/service/impl/BookServiceImpl.java [hunk 12]
```diff
@@ -33,9 +33,25 @@ public List<BookDto> findAll() {
...
-        Book book = bookRepository.findById(id).orElseThrow(() ->
+        Book book = getBookById(id);
+        return bookMapper.toDto(book);
+    }
+
+    @Override
+    public BookDto updateBookById(Long id, CreateBookRequestDto requestDto) {
+        Book book = getBookById(id);
+        Book updatedBook = bookMapper.updateBookFromDto(requestDto, book);
+        return bookMapper.toDto(bookRepository.save(updatedBook));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/service/impl/BookServiceImpl.java]:
```suggestion
        bookMapper.updateBookFromDto(requestDto, book);
        return bookMapper.toDto(bookRepository.save(book));
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Sorry for misunderstanding, let's add DuplicateResourceException and check if isbn exists in service during save book

**Review by sarakhmen — APPROVED** [review]

**Pull request merged**